### DISCUSSION TITLE
単体テストの考え方/使い方 3.6〜4.1.2くらい

### DISCUSSION THREAD

**Discussion by dameleon** [atman-inc/training#96] [General]
- 3.6: 確認(assert)フェーズの読みやすさの改善
- 4.1: 良い単体テストを構成する4本の柱 
    - 4.1.1: 1本目の柱 - 退行(regression)に対する保護
    - 4.1.2: 2本目の柱 - リファクタリングへの耐性

上記辺りののディスカッション。
不明点、相談点、確認点、勉強になったポイントなどは各自ディスカッションのコメントをお願いします。

**Comment by y-kaneda** [comment]
## 3.6 確認(Assert)フェーズの読みやすさの改善
- Assertのライブラリで自然言語のような確認形式にできる
  - BDDだ

## 4.1 良い単体テストを構成する4本の柱
1. 退行(regression)に対する保護
2. リファクタリングへの耐性
3. 迅速なフィードバック
4. 保守のしやすさ
- _感想：基本的にはプロダクションコードと変わらないはず_

### 4.1.1 退行(regression)に対する保護
- 退行は新しい機能が増えるほど、発生する
- **コードは資産ではなく負債**
- 退行を防ぐためのポイントとして以下の3つを意識する
  - テスト時に実行されるプロダクション・コードの量
    - _感想：カバレッジと近い感じかも_
  - そのコードの複雑さ
    - _感想：分岐網羅・条件網羅とか_
  - このコードが扱っているドメインの重要性
    - 要件・仕様に深く関わってくるコードは複雑で、かつ退行によるダメージが非常に大きい
    - 例えば setter/getter みたいな内容の無い処理、静的なコードなどはテストをする価値が低い
- ライブラリやフレームワーク等の外部のコードも気にする必要がある

### 4.1.2 リファクタリングへの耐性
- プロダクションコードへのリファクタリングを行っただけで通らなくなってしまうテストは良くない
  - プロダクションコードが正しいのにテストが失敗することを **偽陽性(false positive)** と呼ぶ
  - 偽陽性があると何がおかしいのか判断が難しくなり、テストに対する信頼性・価値が落ちてしまう。
    - 本来の目的であるプロダクションコードの問題発見ができなくなる(プロダクションコードの問題なのかが分からなくなる)
    - リファクタリングをするのを恐れるようになってしまう
  - _感想：これはテストの観点がおかしいというのがありそう。テスト対象のプロダクションコードの価値を見誤っていそう。コードの振る舞いというよりはその振る舞いの結果を見るべき_

**Comment by xxManatoxx** [comment]
## 3.6: 確認(assert)フェーズの読みやすさの改善

- テストの読みやすさを向上させるために出来ること
  - Assertionのライブラリ導入
  - .NETでは「Fluent Assertions」がある

- 勉強になったポイント
    - 自分が使用している言語でもライブラリを調べてみようと思う

# 第4章

- 第1章で見てきた優れた単体テストの特徴
  - テストすることが開発サイクルの中に組み込まれている
  - コードベースの特に重要な部分のみがテスト対象となっている
  - 最小限の保守コストで最大限の価値を生み出すようになっている

- 価値あるテストを認識することと、作成することは異なる

## 4.1: 良い単体テストを構成する4本の柱

- 4本の柱
  - 退行（Regression）に対する保護
  - リファクタリングへの耐性
  - 迅速なフィードバック
  - 保守のしやすさ

### 4.1.1: 1本目の柱 - 退行(regression)に対する保護

- ソフトウェアで発生する退行とはバグのこと
  - 何らかの変更を加えた後に既存の機能が正常に動かなくなること
- コードは資産ではなく負債

- 退行に対する保護がどのくらい備わっているか
  - テスト時に実行されるプロダクションコードの量
  - そのコードの複雑さ
  - そのコードが扱っているドメインの重要性

### 4.1.2: 2本目の柱 - リファクタリングへの耐性

- テストが失敗すること無く、どのくらいプロダクションコードのリファクタリングを行えるか

- 例：リファクタリングした際にテストが動かなくなったが、既存の振る舞いと変わらずプロダクションコードに問題は見つけられなかった
  - テストコードでプロダクションコードが少しでも変わると失敗するようになっていた
  - テストコードが機能の正しい振る舞いをテストできていなかった

- 上記テストの失敗を偽陽性という
  - 嘘の警告
  - この偽陽性が生まれづらい性質のことを本書では「リファクタリングへの耐性」と呼ぶ

- 気になった点
  - リファクタリングの耐性でどういう実装をしたら振る舞いが同じでもプロダクションコードが少しでも変わると失敗するようになるのかイメージができなかった

**Comment by myamamoto111** [comment]
## 3.6

- 確認(Assert)について
  - 情報をシンプルな物語のように記述

# 第２部　単体テストとその価値
## 第４章　良い単体テストを構成する４本の柱
- **良い単体テストを構成する性質の大きな分類**
- **理想的なテストについての定義**
- **テスト・ピラミッドについて**
- **ブラックボックステストとホワイトボックステストの使い方**

## 4.1
- 良い単体テストを構成するもの
  - 退行(regression)に対する保護
  - リファクタリングへの耐性
  - 迅速なフィードバック
  - 保守のしやすさ

### 4.1.1　退行(regression)に対する保護
- コードは資産ではなく、負債！？
- アプリケーションが使っているライブラリやフレームワークなどの外部から持ち込んだコードも考慮
  - アプリケーションのために書かれたコードと同じくらい機能に影響を与える
- 外部からの依存に対しても想定通りに振る舞っているか検証する

### 4.1.2　リファクタリングへの耐性
- 偽陽性の発生を少なく
- 質の悪いテストケースを取り除く

**Comment by nmurakami0** [comment]
3.6

- BDDのspecに近いかたちでassertを書くと読みやすいという話

4.1

- regressionに対する保護
  - 弊社ではよくデグレと言っている部分
  - "コードが資産ではなく、負債"
  - 保護をテストでどこまで出来ているかの把握観点 (下2つは保護によって生じる価値を表している)
    - テスト時に実行されるプロダクションコードの量
    - そのコードの複雑さ
    - そのコードが扱っているドメインの重要性
- refactoringへの体制
  - 振る舞いをテストすることの重要性がここで出てくる
    - 振る舞いをちゃんとテストしていないと偽陽性がより発生しやすくなる
    - テストケースをテスト対象の内部的なコードから切り離す
  - ただ、渡すパラメータが増えるなど、簡単に治せる偽陽性は許容してよい
  - 監視もそうだが、オオカミ少年的なのは無視していく流れになるのはそれはそう
    - それで重要なアラート見逃すというリスクも同じ

**Comment by koushisa** [comment]
## 3.6

- 主語 -> 動詞 -> 目的語は理解しやすい
  - よく分かる
  - [SVO型](https://ja.wikipedia.org/wiki/SVO%E5%9E%8B)というか、オブジェクト指向の考え方とちかい

## 4

### 4.1.1

- 超大事な言葉
  - 「コードは資産ではなく負債」
  - 「重要なのは、プロダクトの核となる部分（ドメインモデル）」

### 4.1.2

#### false positive

- FE(というかSPA)の状態を扱うプログラミングにおいてこれはとても大きな悩みのタネ
- 思考実験がてら、下記にfalse positiveを避けるための自分の考えを連ねてみる ( !== 読んでもらう文章)
  - 税計算など純粋関数化しやすいものについては言及の対象外

### false positiveとFEのテストについて

#### 結論まとめ

- 時間は有限なので少ないテストケースでより多くのカバレッジを目指しつつたまに探索的テストを書く
- 少ないテストケース
  - FEではブラウザ上でのインタラクション前後のDOMを検証する
  - 本のメッセージレンダラの例でhtmlを検証するのと同じ
  - テストケース定義には必要に応じてステークホルダを巻き込む
  - ライブラリの内側で担保されているような挙動や外部依存はモックで割り切る
- 多くのカバレッジ
  - インタラクションにより初回レンダリングと2回目以降のレンダリングの状態変化を網羅する
  - レンダリングで観測できない内部状態については探索的テストを書く
- [探索的テスト](https://e-words.jp/w/%E6%8E%A2%E7%B4%A2%E7%9A%84%E3%83%86%E3%82%B9%E3%83%88.html)
  - 外部依存や手続き的なコード(=実装詳細)とプロダクトの核となる部分（=ドメインモデル）や純粋関数を積極的に切り離す
  - 状態を扱う部分とそうではない部分を分離して、ロジック部分だけ集中してテストを書く
  - 言語やフレームワーク固有の[制御の反転](https://ja.wikipedia.org/wiki/%E5%88%B6%E5%BE%A1%E3%81%AE%E5%8F%8D%E8%BB%A2)の方法論が頭にあると書きやすい
  - 探索的テストを書くには知識と経験が要されるので、現実は難しい部分も多い  
    - テストすべき箇所とそうではない境界線の見極め
    - 依存制御の設計論的な話や実装分割のテクニカルな知識
    - シンプルな一覧検索CRUDであればドメインモデルなど考えずにhooksへ切り出しつつテストしたい関数だけ分離する、で十分な場面もある
    - 実装がテストに引きずられると目的置換が発生して無限に時間を奪われる

以下は雑多な思考ログ

---

##### 前提のメンタルモデル

- フロントエンドの戦場はたくさんある(toC, UGC, toB)
  - Jamstackな戦場
  - LighthouseやCore Web Vitalsをチューニングする戦場
  - 弊社の主戦場はtoB分野とSPA
    - 複雑な状態遷移を伴う業務システムが多く、Reactが選択されやすい
- システム開発需要(主観)
  - `実装 < 設計`
    - 現実世界の関わる業務プロセスから制約事項を分解してシステム上の仕様に落としこむプロセス
    - 価値、機能、要求に沿ったデザインを実装難易度と天秤にかけて多角的に捉える
    - 対話と信頼関係などによる発展が主で供給も少なく、業務委託で代替するのが難しい
  - `堅牢性 < 開発効率`
    - `QCD`でDは大体トップ
    - 既存の業務プロセスで困っているので、早く解決するために最短でリリースしたいという要望が多い
    - 設計がうまくハマれば実装については業務委託や分業で代替可能ではある
  - `設計 < デザイン`
    - 情報設計とその階層構造や見せ方次第で設計, 実装, 開発効率は全て変わる
    - デザインから暗黙的に設計パターンが決まる
  - `機能の複雑さ = ユーザーにとっての認知負荷`
	- 複雑な機能はユーザーにとっても使いづらいし実装も複雑
	- 複雑な実装方法を考えるよりも機能をシンプルに削ぎ落とすことをリソースを割きたい
- SPA(UI)開発とテスト
  - `テストコードを潤沢にする時間 < 機能をシンプルにできないか考える時間`
	- 機能をシンプルにするとテストケースも減らせる
	- (理想や主観的な議論になりがちで規模が大きく関わる人も多いと合意形成で難しい局面も多い)
  - `実装の正しさ !== ユーザーにとっての価値`
	- フロントエンドはバックエンドと比較すると「価値」の定義が難しい
	- UIに近い位置ほど常に"仮説検証"であることを念頭に置く
  - 定期的に価値そのものが見直される
    - 頑張って作ったけど需要なかった、は割とよくある(システム開発でこのリスクは避けられない)
    - FEはそのようなユーザーの利便性やリブランディングなどの影響をダイレクトに受ける
    - 既存実装は全て捨てて0からやり直す
  - SPAを採用した理由と達成したい目標(ユーザーにとっての価値)を決め、戦略を立て、戦略に合った戦術を選択する  
    - 目標と戦略が間違っていないかを検証するフィードバックループを早く回す方に舵を切るほうが合理的
  - SPAの実装は複雑化になりがちだからこそ戦術(実装&テスト)はミニマムにしたい
    - この文脈でミニマムなテストを考える
    - テストを書くとしたら絞りたい
    - 基本的にBDDっぽいテストになる

##### コンポーネントpropsの組み合わせテストやhooksのテストは書かない

- じゃあ何をテストしたいのかについては後述
- 理由: チーム内で一貫したコンポーネント(問題空間の分割)の粒度を合わせるコストが難しすぎる
  - リファクタリング, 機能追加時にコンポーネントツリーが変わった際にコケまくる
  - 他の人が書いたテストコードがコケた際、問題空間の粒度が合っていないので捨てられやすい
- 自分にとってのコンポーネントとhooksの概念
  - コンポーネント = 文書構造の一部を切り取りマークアップ、スタイリング、アニメーション、ロジックを内包したもの
  - hooksはロジックに当たるものでprivateメソッドとして捉えている
  - compositionや関心の分離のためのテクニックの一つであって主ではない
  - テストにおいてはユーザーの発するイベントを元にUIがどう振る舞うかを検証したい
- 実装をミスってない「安心感」のためにこれらのテストを真面目に書くと開発の見積もりが5割マシぐらいになる
  - 労力に見合った価値を生むか微妙
- 文書構造はデザインとも密接
    - デザインが変わる = 文書構造が変わる = コンポーネント設計が変わる
    - 変更や機能追加の際は前提条件や価値そのものが見直される
    - コンポーネント依存のテストは旨味を感じないケースが経験上多い
- middlewareや責務が一つだけの小さなライブラリ等であればホワイトボックス, パフォーマンステストも必要に応じて書くと思う

##### FEでテストするとしたら

- E2E寄りのテスト
  - 画面仕様(ユーザーが気にする振る舞い)を重視してテストする
  - [StorybookのInteraction tests](https://storybook.js.org/docs/react/writing-tests/interaction-testing)
  -  hooksのテストは状態を目で確認できず、なんのテストをしているのかが非常に分かりづらい
  - Interaction testsはブラウザ上でレンダリング & 操作をエミュレートしてDOMのスナップショットをテストする
    - テストの実行前、中、前後を1ステップずつStorybook上で視認できる
  - インタラクションを軸にテストしてたら初回レンダリングと2回目以降のレンダリングも大体網羅的にテストできる   
  - cypress等と比べてコンポーネントスコープに出来るのと、ロケーターが書きやすい
  - 外部依存はモックする(=フロントエンドで完結する)のでE2Eと結合の中間みたいなイメージ
  - 実行環境はiframeで分離されるのでプロセス間依存も避けられる
- 複雑なロジックや状態遷移を扱う場合
  - 大前提: 当該ロジックをコンポーネントから剥がす
  - 状態遷移
    - 上述のinteraction testsでインタラクション前後のDOMを検証する
  - ロジック
     - hooksの内部処理をドメインモデルに切り出す(pureなtsにする)、ドメインモデル単体に対してテストする
     - フォームのバリデーションスキーマや相関チェック(yup, zod)も関数に切り出してテストする
- 共通コンポーネント
  - 後述のVRTを利用するなら用意するかも？な程度
- a11yテスト
  - プロダクトの特性による
  - toC向けで老若男女使えるWEB標準的な品質を求められるなら欲しい
  - 業務システムだと出番は少ない
  - もしやるならデザインや情報設計のレイヤから入っていかないと厳しい
  - testing-library使ってたら自然とa11yに則した検証ができるケースが多い

##### FEのテストケース

- SPAは組み合わせ爆発を起こしておりワークするFEのテストケース(false positiveが少なく必要なときにコケるテスト)をつくるのは非常に難しい
  - ここに上述の「変更や機能追加の際は前提条件や価値そのものが見直される」も加わる
- デザインや外部仕様にはシステムの内部状態は考慮されないため探索的になる
  - 探索的な箇所はドメインモデルに閉じ込める
- @koushisaの考えるFEのテストケース
  - ユーザーストーリーとある程度地続きな形にしたい
  - 1. 高水準: ユーザーストーリー
  - 2. 中水準:ストーリー内のプロセス間で生じるモジュールの内部状態変化を考慮したインタラクションテスト
  - 3. 低水準: ピュアなドメインモデルに対する探索的テスト
- 中水準のテストを行うためのActが多い場合、複雑な機能(デザイン, 仕様)になっていないかを疑う
- テストフェーズから外部仕様に立ち返ることも多いので、運用や効率を考えると各ステークホルダ(BIZ,DEV,QA)で同じ目線でテストケースをメンテできると良さそうではある
  - フロントエンドは自己完結しない
  - フロントエンドのテストを書くために必要な情報がフロントエンド単体で完結するのは稀
  - [mabl](https://www.mabl.com/ja/)のようなプラットフォームとか模索中

##### FEのテスト容易性の考え方について

- 全体的にコンポーネント設計とテスト容易性は分けて考える
- 非同期処理やモジュールのモックはStorybookのaddonやjestのモック機構がある
- ロジック分割
  - テストの目的により当該の問題空間のみをピュアな関数にする
- コンポーネント分割
  - Container/Presentation
    - コンポーネントのpropsのテストを書くなら分ける
    - 但しコンポーネントツリーとロジックが密結合で壊れやすいテストにはなる
    - 自分は却って凝集度が低く感じるのとコンポーネントのpropsテストのメリットを感じないのであまりやらない
  - ストア層の状態を集約したコンポーネント
    - 実装都合上の責務分割のために作ることはある
    - MVCでいう"Controller"的な役割のコンポーネント
    - 交通整理が役割なのでロジックは持たせない
- 3行で
  - hooksを中心にコンポーネントから重要な状態とロジックを剥がす
  - テストしたいなら関数分割する
  - モックの足回りは整っているのでテストできないなら責務が大きすぎるか何かが間違ってる

##### 非同期処理や外部依存のテスト

- これもStorybookのaddonでモックしてインタラクションテストで担保する
- [msw](https://mswjs.io/)のレイヤで色々できるが現実はFEサイドのAPI理解や習熟度に依存することが多い
  - APIレスポンスのパターンを網羅してテストするのは理想ではあるが現実は厳しい
- インタラクションテストを可能にするためだと割り切ってモックAPIはよしなにでもいいかも
  - 実装する上でAPIの背後にあるリソースの状態や関係性を意識しなくてよい、というわけではない
  - モックに労力を割くコスパと天秤にかけるという意味
  - 長期的には型からAIが生成してくれる世界観が来るはず
- ローディング待ちなどはtesting-libraryとか[Playwright](https://playwright.dev/)の`waitFor`使えばある程度透過的にテストできる
  - `waitFor` = 対象のDOMがレンダリングされるまで内部でポーリングしてくれる

---
以下蛇足だがFEのテスト戦略考えるときに頭の片隅で考慮している変数

##### 番外編1:VRT(Visual Regression Test)

- デグレ基準の共通認識やしきい値の調整などの運用が大変そうなイメージ
  - 気をつけないと大量のfalse positiveが出てくる
  - VRT結果を見ていって問題ないか逐次判断する業務が生まれてしまうのが個人的に苦手
- 自前運用しだすと労力に対してROI上回るかが微妙なので外部サービス使いたい
  - [Chromatic](https://www.chromatic.com/)とか
- デザインシステムと同様、かなりコストがかかるのでやるとしたらビジネスとデザインと開発チームが密にコミュニケーションできる土台づくりから

##### 番外編2: フロントエンドの式年遷宮

- 最近のフレームワーク(Next 13やSvelte Kit)はクラサバを統合してコンポーネントに外部依存を解決できる層を追加した
  - ReactのコアチームやVercelは特にその傾向が強い
  - コンポーネントは文書構造のスライスからDB(リソース)のスライスへ変わろうとしている
- 業務システムでクラサバを統合したフロントエンドを採用した事例が少ないため分業やテスト、モックがどうなるかは未知数
  - 個人的にクラサバ分離したい派(コミュニケーションパスや認知負荷、専門性的な観点)
  - 戦術の一つとして思考実験はした
  - [React Server Componentsをアーキテクチャとしてどう捉えるか](https://scrapbox.io/koushisa/React_Server_Components%E3%82%92%E3%82%A2%E3%83%BC%E3%82%AD%E3%83%86%E3%82%AF%E3%83%81%E3%83%A3%E3%81%A8%E3%81%97%E3%81%A6%E3%81%A9%E3%81%86%E6%8D%89%E3%81%88%E3%82%8B%E3%81%8B)
- 3~5年後にはWEBアプリはGUIやブラウザを脱却してそうな気もする
  - ユーザーはエッジ上にデプロイされたAIをエージェントとして対話しながら自然言語でリソースを操作するみたいな夢物語
  - GUIは時代遅れでVUIと会話型UIが主流みたいな話になるのではと
- 結論: 人間が触るWEBフロントエンドとエントロピー増大は切り離せない
